File: leetcode/simplify-path/simplify-path.py
/**
 * Title: simplify-path
 * Slug: simplify-path
 * URL: https://leetcode.com/problems/simplify-path/?envType=problem-list-v2&envId=string
 * Language: python
 * Submitted: 2025-10-21T18:28:11.638Z
 */

class Solution(object):
    def simplifyPath(self, path):
        """
        :type path: str
        :rtype: str
        """
        stack = []
        for part in path.split('/'):
            if part == '' or part == '.':
                continue
            if part == '..':
                if stack:
                    stack.pop()
            else:
                stack.append(part)
        return '/' + '/'.join(stack)

File: leetcode/simplify-path/simplify-path.java
/**
 * Title: simplify-path
 * Slug: simplify-path
 * URL: https://leetcode.com/problems/simplify-path/?envType=problem-list-v2&envId=string
 * Language: java
 * Submitted: 2025-10-21T18:28:52.634Z
 */


class Solution {
    public String simplifyPath(String path) {
        Deque<String> stack = new ArrayDeque<>();
        for (String part : path.split("/")) {
            if (part.isEmpty() || part.equals(".")) continue;
            if (part.equals("..")) {
                if (!stack.isEmpty()) stack.pollLast();
            } else {
                stack.offerLast(part);
            }
        }
        if (stack.isEmpty()) return "/";
        StringBuilder sb = new StringBuilder();
        for (String s : stack) sb.append('/').append(s);
        return sb.toString();
    }
}